[BUG] test_multi_types_window_aggs_for_rows_lead_lag fails against Spark 3.1.0

**Describe the bug**
Nightly build of integration tests against Spark 3.1.0 are failing in the test_multi_types_window_aggs_for_lead_lag tests.

**Steps/Code to reproduce bug**
Run integration tests against latest Spark 3.1.0-SNAPSHOT

**Expected behavior**
Tests do not fail.

Looks like the GPU is somehow returning nulls or no result.  Here's an excerpt from the failure log.  Note the `gpu = None`.
```
11:24:30  =================================== FAILURES ===================================
11:24:30  _ test_multi_types_window_aggs_for_rows_lead_lag[partBy:Long-orderBy:Long-Long] _
11:24:30  
11:24:30  a_gen = Long, b_gen = Long, c_gen = Long
11:24:30  
11:24:30      @ignore_order
11:24:30      @approximate_float
11:24:30      @pytest.mark.parametrize('c_gen', lead_lag_data_gens, ids=idfn)
11:24:30      @pytest.mark.parametrize('b_gen', part_and_order_gens, ids=meta_idfn('orderBy:'))
11:24:30      @pytest.mark.parametrize('a_gen', part_and_order_gens, ids=meta_idfn('partBy:'))
11:24:30      def test_multi_types_window_aggs_for_rows_lead_lag(a_gen, b_gen, c_gen):
11:24:30          data_gen = [
11:24:30                  ('a', RepeatSeqGen(a_gen, length=20)),
11:24:30                  ('b', b_gen),
11:24:30                  ('c', c_gen)]
11:24:30          # By default for many operations a range of unbounded to unbounded is used
11:24:30          # This will not work until https://github.com/NVIDIA/spark-rapids/issues/216
11:24:30          # is fixed.
11:24:30      
11:24:30          # Ordering needs to include c because with nulls and especially on booleans
11:24:30          # it is possible to get a different ordering when it is ambiguous.
11:24:30          baseWindowSpec = Window.partitionBy('a').orderBy('b', 'c')
11:24:30          inclusiveWindowSpec = baseWindowSpec.rowsBetween(-10, 100)
11:24:30      
11:24:30          defaultVal = gen_scalar_value(c_gen, force_no_nulls=False)
11:24:30      
11:24:30          def do_it(spark):
11:24:30              return gen_df(spark, data_gen, length=2048) \
11:24:30                      .withColumn('inc_count_1', f.count('*').over(inclusiveWindowSpec)) \
11:24:30                      .withColumn('inc_count_c', f.count('c').over(inclusiveWindowSpec)) \
11:24:30                      .withColumn('inc_max_c', f.max('c').over(inclusiveWindowSpec)) \
11:24:30                      .withColumn('inc_min_c', f.min('c').over(inclusiveWindowSpec)) \
11:24:30                      .withColumn('lead_5_c', f.lead('c', 5).over(baseWindowSpec)) \
11:24:30                      .withColumn('lead_def_c', f.lead('c', 2, defaultVal).over(baseWindowSpec)) \
11:24:30                      .withColumn('lag_1_c', f.lag('c', 1).over(baseWindowSpec)) \
11:24:30                      .withColumn('lag_def_c', f.lag('c', 4, defaultVal).over(baseWindowSpec)) \
11:24:30                      .withColumn('row_num', f.row_number().over(baseWindowSpec))
11:24:30  >       assert_gpu_and_cpu_are_equal_collect(do_it, conf={'spark.rapids.sql.hasNans': 'false'})
11:24:30  
11:24:30  src/main/python/window_function_test.py:113: 
11:24:30  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
11:24:30  src/main/python/asserts.py:296: in assert_gpu_and_cpu_are_equal_collect
11:24:30      _assert_gpu_and_cpu_are_equal(func, True, conf=conf)
11:24:30  src/main/python/asserts.py:288: in _assert_gpu_and_cpu_are_equal
11:24:30      assert_equal(from_cpu, from_gpu)
11:24:30  src/main/python/asserts.py:86: in assert_equal
11:24:30      _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
11:24:30  src/main/python/asserts.py:38: in _assert_equal
11:24:30      _assert_equal(cpu[index], gpu[index], float_check, path + [index])
11:24:30  src/main/python/asserts.py:31: in _assert_equal
11:24:30      _assert_equal(cpu[field], gpu[field], float_check, path + [field])
11:24:30  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
11:24:30  
11:24:30  cpu = -6068833985444302436, gpu = None
11:24:30  float_check = <function get_float_check.<locals>.<lambda> at 0x7f63ca8033b0>
11:24:30  path = [1, 'lag_1_c']
11:24:30  
11:24:30      def _assert_equal(cpu, gpu, float_check, path):
11:24:30          t = type(cpu)
11:24:30          if (t is Row):
11:24:30              assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {}".format(path)
11:24:30              if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):
11:24:30                  for field in cpu.__fields__:
11:24:30                      _assert_equal(cpu[field], gpu[field], float_check, path + [field])
11:24:30              else:
11:24:30                  for index in range(len(cpu)):
11:24:30                      _assert_equal(cpu[index], gpu[index], float_check, path + [index])
11:24:30          elif (t is list):
11:24:30              assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {}".format(path)
11:24:30              for index in range(len(cpu)):
11:24:30                  _assert_equal(cpu[index], gpu[index], float_check, path + [index])
11:24:30          elif (t is pytypes.GeneratorType):
11:24:30              index = 0
11:24:30              # generator has no zip :( so we have to do this the hard way
11:24:30              done = False
11:24:30              while not done:
11:24:30                  sub_cpu = None
11:24:30                  sub_gpu = None
11:24:30                  try:
11:24:30                      sub_cpu = next(cpu)
11:24:30                  except StopIteration:
11:24:30                      done = True
11:24:30      
11:24:30                  try:
11:24:30                      sub_gpu = next(gpu)
11:24:30                  except StopIteration:
11:24:30                      done = True
11:24:30      
11:24:30                  if done:
11:24:30                      assert sub_cpu == sub_gpu and sub_cpu == None, "CPU and GPU generators have different lengths at {}".format(path)
11:24:30                  else:
11:24:30                      _assert_equal(sub_cpu, sub_gpu, float_check, path + [index])
11:24:30      
11:24:30                  index = index + 1
11:24:30          elif (t is int):
11:24:30  >           assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)
11:24:30  E           AssertionError: GPU and CPU int values are different at [1, 'lag_1_c']
11:24:30  
11:24:30  src/main/python/asserts.py:63: AssertionError
11:24:30  ----------------------------- Captured stdout call -----------------------------
11:24:30  ### CPU RUN ###
11:24:30  ### GPU RUN ###
11:24:30  ### COLLECT: GPU TOOK 0.8450026512145996 CPU TOOK 0.69598388671875 ###
11:24:30  _ test_multi_types_window_aggs_for_rows_lead_lag[partBy:Long-orderBy:Long-Double] _
11:24:30  
11:24:30  a_gen = Long, b_gen = Long, c_gen = Double
11:24:30  
11:24:30      @ignore_order
11:24:30      @approximate_float
11:24:30      @pytest.mark.parametrize('c_gen', lead_lag_data_gens, ids=idfn)
11:24:30      @pytest.mark.parametrize('b_gen', part_and_order_gens, ids=meta_idfn('orderBy:'))
11:24:30      @pytest.mark.parametrize('a_gen', part_and_order_gens, ids=meta_idfn('partBy:'))
11:24:30      def test_multi_types_window_aggs_for_rows_lead_lag(a_gen, b_gen, c_gen):
11:24:30          data_gen = [
11:24:30                  ('a', RepeatSeqGen(a_gen, length=20)),
11:24:30                  ('b', b_gen),
11:24:30                  ('c', c_gen)]
11:24:30          # By default for many operations a range of unbounded to unbounded is used
11:24:30          # This will not work until https://github.com/NVIDIA/spark-rapids/issues/216
11:24:30          # is fixed.
11:24:30      
11:24:30          # Ordering needs to include c because with nulls and especially on booleans
11:24:30          # it is possible to get a different ordering when it is ambiguous.
11:24:30          baseWindowSpec = Window.partitionBy('a').orderBy('b', 'c')
11:24:30          inclusiveWindowSpec = baseWindowSpec.rowsBetween(-10, 100)
11:24:30      
11:24:30          defaultVal = gen_scalar_value(c_gen, force_no_nulls=False)
11:24:30      
11:24:30          def do_it(spark):
11:24:30              return gen_df(spark, data_gen, length=2048) \
11:24:30                      .withColumn('inc_count_1', f.count('*').over(inclusiveWindowSpec)) \
11:24:30                      .withColumn('inc_count_c', f.count('c').over(inclusiveWindowSpec)) \
11:24:30                      .withColumn('inc_max_c', f.max('c').over(inclusiveWindowSpec)) \
11:24:30                      .withColumn('inc_min_c', f.min('c').over(inclusiveWindowSpec)) \
11:24:30                      .withColumn('lead_5_c', f.lead('c', 5).over(baseWindowSpec)) \
11:24:30                      .withColumn('lead_def_c', f.lead('c', 2, defaultVal).over(baseWindowSpec)) \
11:24:30                      .withColumn('lag_1_c', f.lag('c', 1).over(baseWindowSpec)) \
11:24:30                      .withColumn('lag_def_c', f.lag('c', 4, defaultVal).over(baseWindowSpec)) \
11:24:30                      .withColumn('row_num', f.row_number().over(baseWindowSpec))
11:24:30  >       assert_gpu_and_cpu_are_equal_collect(do_it, conf={'spark.rapids.sql.hasNans': 'false'})
11:24:30  
11:24:30  src/main/python/window_function_test.py:113: 
11:24:30  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
11:24:30  src/main/python/asserts.py:296: in assert_gpu_and_cpu_are_equal_collect
11:24:30      _assert_gpu_and_cpu_are_equal(func, True, conf=conf)
11:24:30  src/main/python/asserts.py:288: in _assert_gpu_and_cpu_are_equal
11:24:30      assert_equal(from_cpu, from_gpu)
11:24:30  src/main/python/asserts.py:86: in assert_equal
11:24:30      _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
11:24:30  src/main/python/asserts.py:38: in _assert_equal
11:24:30      _assert_equal(cpu[index], gpu[index], float_check, path + [index])
11:24:30  src/main/python/asserts.py:31: in _assert_equal
11:24:30      _assert_equal(cpu[field], gpu[field], float_check, path + [field])
11:24:30  src/main/python/asserts.py:68: in _assert_equal
11:24:30      assert float_check(cpu, gpu), "GPU and CPU float values are different {}".format(path)
11:24:30  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
11:24:30  
11:24:30  lhs = -3.546714417938851e-160, rhs = None
11:24:30  
11:24:30  >   return lambda lhs,rhs: lhs == pytest.approx(rhs, **_approximate_float_args)
11:24:30  E   TypeError: cannot make approximate comparisons to non-numeric values: None
11:24:30  
11:24:30  src/main/python/conftest.py:24: TypeError
```

--- a/integration_tests/src/main/python/window_function_test.py
+++ b/integration_tests/src/main/python/window_function_test.py
@@ -101,7 +101,6 @@ def meta_idfn(meta):
         return meta + idfn(something)
     return tmp
 
-@pytest.mark.xfail(condition=not(is_before_spark_310()), reason='https://github.com/NVIDIA/spark-rapids/issues/999')
 @ignore_order
 @approximate_float
 @pytest.mark.parametrize('c_gen', lead_lag_data_gens, ids=idfn)

--- a/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/OffsetWindowFunctionMeta.scala
+++ b/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/OffsetWindowFunctionMeta.scala
@@ -37,9 +37,8 @@ abstract class OffsetWindowFunctionMeta[INPUT <: OffsetWindowFunction] (
     parent: Option[RapidsMeta[_, _, _]],
     rule: DataFromReplacementRule)
   extends ExprMeta[INPUT](expr, conf, parent, rule) {
-  val input: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.input, conf, Some(this))
-  // For LAG() expressions, negate the offset value.
-  val adjustedOffset: Expression = {
+  lazy val input: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.input, conf, Some(this))
+  lazy val adjustedOffset: Expression = {
     expr match {
       case lag: Lag =>
         GpuOverrides.extractLit(lag.offset) match {
@@ -61,8 +60,23 @@ abstract class OffsetWindowFunctionMeta[INPUT <: OffsetWindowFunction] (
         throw new IllegalStateException(s"$other is not a supported window function")
     }
   }
-  val offset: BaseExprMeta[_] =
+  lazy val offset: BaseExprMeta[_] =
     GpuOverrides.wrapExpr(adjustedOffset, conf, Some(this))
-  val default: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.default, conf, Some(this))
-  override val childExprs: Seq[BaseExprMeta[_]] = Seq(input, offset, default)
+  lazy val default: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.default, conf, Some(this))
+
+  override val childExprs: Seq[BaseExprMeta[_]] = Seq.empty
+
+  override def tagExprForGpu(): Unit = {
+    expr match {
+      case Lead(_,_,_) => // Supported.
+      case Lag(_,_,_) =>  // Supported.
+      case other =>
+        willNotWorkOnGpu( s"Only LEAD/LAG offset window functions are supported. Found: $other")
+    }
+
+    if (GpuOverrides.extractLit(expr.offset).isEmpty) { // Not a literal offset.
+      willNotWorkOnGpu(
+        s"Only integer literal offsets are supported for LEAD/LAG. Found: ${expr.offset}")
+    }
+  }
 }

--- a/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/OffsetWindowFunctionMeta.scala
+++ b/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/OffsetWindowFunctionMeta.scala
@@ -43,7 +43,7 @@ abstract class OffsetWindowFunctionMeta[INPUT <: OffsetWindowFunction] (
     expr match {
       case lag: Lag =>
         GpuOverrides.extractLit(lag.offset) match {
-         case Some(Literal(offset:Int, IntegerType)) =>
+         case Some(Literal(offset: Int, IntegerType)) =>
             Literal(-offset, IntegerType)
          case _ =>
            throw new IllegalStateException(

--- a/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/OffsetWindowFunctionMeta.scala
+++ b/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/OffsetWindowFunctionMeta.scala
@@ -1,0 +1,57 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.shims.spark311
+
+import com.nvidia.spark.rapids.{BaseExprMeta, DataFromReplacementRule, ExprMeta, GpuOverrides, RapidsConf, RapidsMeta}
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, Lag, Literal, OffsetWindowFunction}
+import org.apache.spark.sql.types.IntegerType
+
+/**
+ * Spark 3.1.1-specific replacement for com.nvidia.spark.rapids.OffsetWindowFunctionMeta.
+ * This is required primarily for two reasons:
+ *   1. com.nvidia.spark.rapids.OffsetWindowFunctionMeta (compiled against Spark 3.0.x)
+ *      fails class load in Spark 3.1.x. (`expr.input` is not recognized as an Expression.)
+ *   2. The semantics of offsets in LAG() are reversed/negated in Spark 3.1.1.
+ *      E.g. The expression `LAG(col, 5)` causes Lag.offset to be set to `-5`,
+ *      as opposed to `5`, in prior versions of Spark.
+ * This class adjusts the LAG offset to use similar semantics to Spark 3.0.x.
+ */
+abstract class OffsetWindowFunctionMeta[INPUT <: OffsetWindowFunction] (
+    expr: INPUT,
+    conf: RapidsConf,
+    parent: Option[RapidsMeta[_, _, _]],
+    rule: DataFromReplacementRule)
+  extends ExprMeta[INPUT](expr, conf, parent, rule) {
+  val input: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.input, conf, Some(this))
+  // For LAG() expressions, negate the offset value.
+  val adjustedOffset: Expression = {
+    expr match {
+      case lag: Lag =>
+        lag.offset match {
+          case Literal(v: Int, IntegerType) =>
+            Literal(-v)
+          case _ => expr.offset
+        }
+      case _ => expr.offset
+    }
+  }
+  val offset: BaseExprMeta[_] =
+    GpuOverrides.wrapExpr(adjustedOffset, conf, Some(this))
+  val default: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.default, conf, Some(this))
+  override val childExprs: Seq[BaseExprMeta[_]] = Seq(input, offset, default)
+}

--- a/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/OffsetWindowFunctionMeta.scala
+++ b/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/OffsetWindowFunctionMeta.scala
@@ -18,7 +18,7 @@ package com.nvidia.spark.rapids.shims.spark311
 
 import com.nvidia.spark.rapids.{BaseExprMeta, DataFromReplacementRule, ExprMeta, GpuOverrides, RapidsConf, RapidsMeta}
 
-import org.apache.spark.sql.catalyst.expressions.{Expression, Lag, Literal, OffsetWindowFunction}
+import org.apache.spark.sql.catalyst.expressions.{Expression, Lag, Lead, Literal, OffsetWindowFunction}
 import org.apache.spark.sql.types.IntegerType
 
 /**
@@ -42,12 +42,23 @@ abstract class OffsetWindowFunctionMeta[INPUT <: OffsetWindowFunction] (
   val adjustedOffset: Expression = {
     expr match {
       case lag: Lag =>
-        lag.offset match {
-          case Literal(v: Int, IntegerType) =>
-            Literal(-v)
-          case _ => expr.offset
+        GpuOverrides.extractLit(lag.offset) match {
+         case Some(Literal(offset:Int, IntegerType)) =>
+            Literal(-offset, IntegerType)
+         case _ =>
+           throw new IllegalStateException(
+             s"Only integer literal offsets are supported for LAG. Found:${lag.offset}")
         }
-      case _ => expr.offset
+      case lead: Lead =>
+        GpuOverrides.extractLit(lead.offset) match {
+          case Some(Literal(offset: Int, IntegerType)) =>
+            Literal(offset, IntegerType)
+          case _ =>
+            throw new IllegalStateException(
+              s"Only integer literal offsets are supported for LEAD. Found:${lead.offset}")
+        }
+      case other =>
+        throw new IllegalStateException(s"$other is not a supported window function")
     }
   }
   val offset: BaseExprMeta[_] =

--- a/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala
+++ b/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala
@@ -45,6 +45,21 @@ import org.apache.spark.sql.rapids.shims.spark311._
 import org.apache.spark.sql.types._
 import org.apache.spark.storage.{BlockId, BlockManagerId}
 
+abstract class OffsetWindowFunctionMeta2[INPUT <: OffsetWindowFunction] (
+    isLag: Boolean,
+    expr: INPUT,
+    conf: RapidsConf,
+    parent: Option[RapidsMeta[_, _, _]],
+    rule: DataFromReplacementRule)
+  extends ExprMeta[INPUT](expr, conf, parent, rule) {
+  val input: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.input, conf, Some(this))
+  val adjustedOffset: Expression = if(isLag) UnaryMinus(expr.offset) else expr.offset
+  val offset: BaseExprMeta[_] =
+    GpuOverrides.wrapExpr(adjustedOffset, conf, Some(this))
+  val default: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.default, conf, Some(this))
+  override val childExprs: Seq[BaseExprMeta[_]] = Seq(input, offset, default)
+}
+
 class Spark311Shims extends Spark301Shims {
 
   override def getSparkShimVersion: ShimVersion = SparkShimServiceProvider.VERSION
@@ -138,7 +153,6 @@ class Spark311Shims extends Spark301Shims {
 
         // stringChecks are the same
         // binaryChecks are the same
-
         override val decimalChecks: TypeSig = none
         override val sparkDecimalSig: TypeSig = numeric + BOOLEAN + STRING
 
@@ -196,8 +210,37 @@ class Spark311Shims extends Spark301Shims {
             childExprs(1).convertToGpu(),
             childExprs(2).convertToGpu())
         }
+      }),
+    // Spark 3.1.1-specific LEAD expression, using custom OffsetWindowFunctionMeta.
+    GpuOverrides.expr[Lead](
+      "Window function that returns N entries ahead of this one",
+      ExprChecks.windowOnly(TypeSig.numeric + TypeSig.BOOLEAN +
+        TypeSig.DATE + TypeSig.TIMESTAMP, TypeSig.all,
+        Seq(ParamCheck("input", TypeSig.numeric + TypeSig.BOOLEAN +
+          TypeSig.DATE + TypeSig.TIMESTAMP, TypeSig.all),
+          ParamCheck("offset", TypeSig.INT, TypeSig.INT),
+          ParamCheck("default", TypeSig.numeric + TypeSig.BOOLEAN +
+            TypeSig.DATE + TypeSig.TIMESTAMP + TypeSig.NULL, TypeSig.all))),
+      (lead, conf, p, r) => new OffsetWindowFunctionMeta[Lead](lead, conf, p, r) {
+        override def convertToGpu(): GpuExpression =
+          GpuLead(input.convertToGpu(), offset.convertToGpu(), default.convertToGpu())
+      }),
+    // Spark 3.1.1-specific LAG expression, using custom OffsetWindowFunctionMeta.
+    GpuOverrides.expr[Lag](
+      "Window function that returns N entries behind this one",
+      ExprChecks.windowOnly(TypeSig.numeric + TypeSig.BOOLEAN +
+        TypeSig.DATE + TypeSig.TIMESTAMP, TypeSig.all,
+        Seq(ParamCheck("input", TypeSig.numeric + TypeSig.BOOLEAN +
+          TypeSig.DATE + TypeSig.TIMESTAMP, TypeSig.all),
+          ParamCheck("offset", TypeSig.INT, TypeSig.INT),
+          ParamCheck("default", TypeSig.numeric + TypeSig.BOOLEAN +
+            TypeSig.DATE + TypeSig.TIMESTAMP + TypeSig.NULL, TypeSig.all))),
+      (lag, conf, p, r) => new OffsetWindowFunctionMeta[Lag](lag, conf, p, r) {
+        override def convertToGpu(): GpuExpression = {
+          GpuLag(input.convertToGpu(), offset.convertToGpu(), default.convertToGpu())
+        }
       })
-  ).map(r => (r.getClassFor.asSubclass(classOf[Expression]), r)).toMap
+ ).map(r => (r.getClassFor.asSubclass(classOf[Expression]), r)).toMap
 
   override def getExprs: Map[Class[_ <: Expression], ExprRule[_ <: Expression]] = {
     super.exprs301 ++ exprs311

--- a/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala
+++ b/shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala
@@ -45,21 +45,6 @@ import org.apache.spark.sql.rapids.shims.spark311._
 import org.apache.spark.sql.types._
 import org.apache.spark.storage.{BlockId, BlockManagerId}
 
-abstract class OffsetWindowFunctionMeta2[INPUT <: OffsetWindowFunction] (
-    isLag: Boolean,
-    expr: INPUT,
-    conf: RapidsConf,
-    parent: Option[RapidsMeta[_, _, _]],
-    rule: DataFromReplacementRule)
-  extends ExprMeta[INPUT](expr, conf, parent, rule) {
-  val input: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.input, conf, Some(this))
-  val adjustedOffset: Expression = if(isLag) UnaryMinus(expr.offset) else expr.offset
-  val offset: BaseExprMeta[_] =
-    GpuOverrides.wrapExpr(adjustedOffset, conf, Some(this))
-  val default: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.default, conf, Some(this))
-  override val childExprs: Seq[BaseExprMeta[_]] = Seq(input, offset, default)
-}
-
 class Spark311Shims extends Spark301Shims {
 
   override def getSparkShimVersion: ShimVersion = SparkShimServiceProvider.VERSION

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala
@@ -739,7 +739,25 @@ abstract class OffsetWindowFunctionMeta[INPUT <: OffsetWindowFunction] (
     rule: DataFromReplacementRule)
     extends ExprMeta[INPUT](expr, conf, parent, rule) {
   val input: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.input, conf, Some(this))
-  val offset: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.offset, conf, Some(this))
+  val offset: BaseExprMeta[_] = {
+    expr match {
+      case Lead(_,_,_) => // Supported.
+      case Lag(_,_,_) =>  // Supported.
+      case other =>
+        throw new IllegalStateException(
+          s"Only LEAD/LAG offset window functions are supported. Found: $other")
+    }
+
+    val literalOffset = GpuOverrides.extractLit(expr.offset) match {
+      case Some(Literal(offset: Int, IntegerType)) =>
+        Literal(offset, IntegerType)
+      case _ =>
+        throw new IllegalStateException(
+          s"Only integer literal offsets are supported for LEAD/LAG. Found: ${expr.offset}")
+    }
+
+    GpuOverrides.wrapExpr(literalOffset, conf, Some(this))
+  }
   val default: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.default, conf, Some(this))
   override val childExprs: Seq[BaseExprMeta[_]] = Seq(input, offset, default)
 }

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala
@@ -738,8 +738,8 @@ abstract class OffsetWindowFunctionMeta[INPUT <: OffsetWindowFunction] (
     parent: Option[RapidsMeta[_, _, _]],
     rule: DataFromReplacementRule)
     extends ExprMeta[INPUT](expr, conf, parent, rule) {
-  val input: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.input, conf, Some(this))
-  val offset: BaseExprMeta[_] = {
+  lazy val input: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.input, conf, Some(this))
+  lazy val offset: BaseExprMeta[_] = {
     expr match {
       case Lead(_,_,_) => // Supported.
       case Lag(_,_,_) =>  // Supported.
@@ -758,8 +758,23 @@ abstract class OffsetWindowFunctionMeta[INPUT <: OffsetWindowFunction] (
 
     GpuOverrides.wrapExpr(literalOffset, conf, Some(this))
   }
-  val default: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.default, conf, Some(this))
-  override val childExprs: Seq[BaseExprMeta[_]] = Seq(input, offset, default)
+  lazy val default: BaseExprMeta[_] = GpuOverrides.wrapExpr(expr.default, conf, Some(this))
+
+  override val childExprs: Seq[BaseExprMeta[_]] = Seq.empty
+
+  override def tagExprForGpu(): Unit = {
+    expr match {
+      case Lead(_,_,_) => // Supported.
+      case Lag(_,_,_) =>  // Supported.
+      case other =>
+        willNotWorkOnGpu( s"Only LEAD/LAG offset window functions are supported. Found: $other")
+    }
+
+    if (GpuOverrides.extractLit(expr.offset).isEmpty) { // Not a literal offset.
+      willNotWorkOnGpu(
+        s"Only integer literal offsets are supported for LEAD/LAG. Found: ${expr.offset}")
+    }
+  }
 }
 
 trait GpuOffsetWindowFunction extends GpuAggregateWindowFunction {

